[Order creation] Update coupons display order details

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14097
Ref: pe5pgL-5G4-p2#comment-4599

This PR updates the how coupons are displayed during Order creation. Since we do not have a design proposal here, we just went along copying and adapting the shipping lines UI. 

I modified minimum possible in order to avoid scope creep, but I think we can still extract more logic and encapsulate it outside the `EditableOrderViewModel` better than we're currently doing.

### Changes:
- Removed the ability and views to edit coupons from the Order creation flow. These are either added or deleted.
- As with shipping lines, moved the order coupon observation to its own `EditableOrderCouponLineViewModel`
- As with shipping lines, created a `OrderCouponSectionView`, driven by the view model above.
- No further tests have been added, the existing logic is already tested through `CouponLineDetailsViewModel`, `RemoteOrderSynchronizerTests`, etc, ...

**Before:**
* Coupons can be added from Order details, but do not show, now are editable.
* Coupons show in the drawer, with a pencil to edit/delete them.

**After:**
* Coupons show individually in the Order Details, and can be deleted per row basis.
* Coupons are no longer editable from this view, we either add them or delete them to the Order.
* Coupons show in the drawer as read-only.

| Before | After |
|--------|--------|
| <img width="484" alt="Screenshot 2024-10-03 at 11 22 43" src="https://github.com/user-attachments/assets/d2c7e00a-a28b-482e-8656-f221b3f8ea1e"> | <img width="484" src="https://github.com/user-attachments/assets/d9b0ba0f-eba6-4e21-8a1d-16d6d433e2a7"> | 

## Testing
**Case: Add coupon to order success:**
* Create a new order, observe that the `+ Add Coupon` row is disabled.
* Add a product to the order, observe that the `+ Add Coupon` row is now enabled.
* Tap on `+ Add Coupon` -> Apply a eligible coupon to that product -> Observe how it appears in the Order, with a "trash bin" button.
* Expand the bottom drawer, observe how the coupon appears in the order details.

**Case: Delete coupon from order:**
* Tap on the trash icon and observe how the coupon is removed from the order.

**Case: Add coupon to order error**
* Attempt to add a non-eligible coupon to the order (for example: a coupon that is only applicable to product category A, but the order has a product category B)
* Observe how the coupon is not added to the order, and a error notice appears at the bottom

**Case: Editing an existing order**
* Open an order that has coupons applied
* Observe how the coupons appear as a list, titled as "Discount" (this is pre-existent)
* Tap on `Edit`
* Observe the coupon rows are editable normally

| See existing order | Edit existing order |
|--------|--------|
| <img width="484" src="https://github.com/user-attachments/assets/6c69f9c1-bf49-4655-8810-c9ba43c72561"> | <img width="484" src="https://github.com/user-attachments/assets/d9b0ba0f-eba6-4e21-8a1d-16d6d433e2a7"> | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Remove coupon action from drawer

CouponLineDetails cannot longer be called from the order drawner. This is being moved to the Order details.

iamgabrielma committed:
Delete CouponLineDetails view

This was only used for coupon interaction with the order drawer, now that the feature no longer is there we can remove the view as well.

iamgabrielma committed:
Make view and vm for managing coupons in order

Since coupons needs its own section within the editable order, we make a specific view and vm to manage those.

iamgabrielma committed:
Render coupons section. Pass coupons vm into components section

iamgabrielma committed:
Add order coupon line vm to editable order

iamgabrielma committed:
Linting

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Allow coupon deletion from order creation

iamgabrielma committed:
Copy UI for coupons in order as used in shipping lines

iamgabrielma committed:
Update UI for coupons in drawer

With the UI update, I’ve also moved some logic outside of the the EditableOrder viewmodel, there’s a bunch of things that are currently very coupled with the order and shouldn’t be. This still can be improved, but for scope I’m keeping changes at a minimum.

iamgabrielma committed:
Make tests compile

iamgabrielma commented:
@staskus this can wait post Woo DM for review and iterations, just leaving it in the review queue as I'll be AFK next week 👍

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Coupons/EditableOrderCouponLineViewModel.swift:
```diff
@@ -0,0 +1,22 @@
+        observeCouponsInOrder()
+    }
+
+    func observeCouponsInOrder() {
```
I just realized this can be private, I'll update it before wrapping-up.

staskus reviewed (commented):
Thanks for the work. The functionality works well and in my subjective opinion also allows for better and more consistent UX:

✅ Case: Add coupon to order success
✅ Case: Delete coupon from order:
✅ Case: Add coupon to order error
✅ Case: Editing an existing order

----

A couple of nitpicks:

**1. Inconsistent font styling**

Other rows such as Shipping or Custom Amount use a regular font to display titles:

![Inconsistent styling - names are written in regular font](https://github.com/user-attachments/assets/ea42c426-e7cb-4fa4-a54b-83ba60f5f03e)

**2. Plus button is active but unresponsive**

When the coupon is added and the order is refreshing, "+" button is still active although unresponsive.

![Nitpick - plus button active but unresponsive when adding items](https://github.com/user-attachments/assets/8a9930ce-dc8c-4599-bbf6-ee8e6b862688)

staskus commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Coupons/EditableOrderCouponLineViewModel.swift:
```diff
@@ -0,0 +1,22 @@
+final class EditableOrderCouponLineViewModel: ObservableObject {
+    @Published private(set) var couponLineRows: [OrderCouponLine] = []
+
+    private var orderSynchronizer: OrderSynchronizer
```
```suggestion
    private let orderSynchronizer: OrderSynchronizer
```

iamgabrielma committed:
Merge branch 'trunk' into issue/14097-update-coupons-display-order-details

iamgabrielma committed:
Adjust access control to private. Make let

iamgabrielma commented:
Thanks for the review @staskus , I have a question:

> Other rows such as Shipping or Custom Amount use a regular font to display titles

I cannot see the difference in the titles, all of them seem to use `HeadlineStyle()`? Or are you referring to the `coupon.code` itself? Eg: "13-percentage-discount" and that this one should be showing using regular font rather than "greyed out"? 
Example:
<img width="391" alt="Screenshot 2024-10-21 at 17 03 22" src="https://github.com/user-attachments/assets/1582cd36-dfe6-4065-93a7-eafa1aee0509">

Let me know if I miss-understood, addressed here: https://github.com/woocommerce/woocommerce-ios/pull/14107/commits/71c64ecec1e13e94d5a1e815618fcf1b602c74b0

iamgabrielma committed:
Update plus button style when model is loading

iamgabrielma committed:
update coupon title style to use regular font

iamgabrielma commented:
> When the coupon is added and the order is refreshing, "+" button is still active although unresponsive.

Good catch! Updated so the plus button is disabled and its style also changes while the order is syncing: https://github.com/woocommerce/woocommerce-ios/pull/14107/commits/ad868c3480b8bacc391a7036e1f5ef30686e0392 (maybe we should also grey out the whole row? 🤔 I'd prefer to not invest too much time into the UI in any case for the moment)

![Simulator Screen Recording - iPhone 15 - 2024-10-21 at 17 34 07](https://github.com/user-attachments/assets/b947a0f5-0fa0-406e-83b1-b0096e6de3e5)

Ready for re-review!

staskus reviewed (approved):
>  Or are you referring to the coupon.code itself? Eg: "13-percentage-discount" and that this one should be showing using regular font rather than "greyed out"?

Yes, that's what I meant.

Thank you for changes!
